javaslang/src/main/java/javaslang/collection/PriorityQueue.java

```diff
@@ -0,0 +1,853 @@
+        this.comparator = SerializableComparator.of(comparator);
+        this.forest = forest;
+    }
+
+    abstract Q with(List<Node<PriorityQueue.Ranked<T>>> forest);
+
+    private static <T> T root(Node<Ranked<T>> tree) {
+        return tree.getValue().value;
+    }
+
+    private static <T> int rank(Node<Ranked<T>> tree) {
+        return tree.getValue().rank;
+    }
+
+    /**
+     * fun link (t1 as Node (x1,r1,c1), t2 as Node (x2,r2,c2)) = (∗ r1 = r2 ∗)
+     * *  if Elem.leq (x1,x2) then Node (x1,r1+1,t2 :: c1)
+     * *  else                     Node (x2,r2+1,t1 :: c2
+     */
+    private Node<Ranked<T>> link(Node<Ranked<T>> t1, Node<Ranked<T>> t2) {
+        // TODO refactor
+        final int r1 = rank(t1) + 1, r2 = rank(t2) + 1;
+        final T x1 = root(t1), x2 = root(t2);
+
+        final Ranked<T> ranked;
+        final List<Node<Ranked<T>>> children;
+        if (comparator.isLessOrEqual(x1, x2)) {
+            ranked = Ranked.of(x1, r1);
+            children = List.of(t2).appendAll(t1.getChildren());
+        } else {
+            ranked = Ranked.of(x2, r2);
+            children = List.of(t1).appendAll(t2.getChildren());
+        }
+        return Tree.of(ranked, children);
+    }
+
+    /**
+     * fun skewLink (t0 as Node (x0,r0, _), t1 as Node (x1,r1,c1), t2 as Node (x2,r2,c2)) =
+     * *  if Elem.leq (x1,x0) andalso Elem.leq (x1,x2) then      Node (x1,r1+1,t0 :: t2 :: c1)
+     * *  else if Elem.leq (x2,x0) andalso Elem.leq (x2,x1) then Node (x2,r2+1,t0 :: t1 :: c2)
+     * *  else                                                   Node (x0,r1+1,[t1, t2])
+     **/
+    private Node<Ranked<T>> skewLink(Node<Ranked<T>> t0, Node<Ranked<T>> t1, Node<Ranked<T>> t2) {
+        // TODO refactor
+        final T x0 = root(t0);
+
+        final int r1 = rank(t1) + 1, r2 = rank(t2) + 1;
+        final T x1 = root(t1), x2 = root(t2);
+
+        final Ranked<T> ranked;
+        final List<Node<Ranked<T>>> children;
+        if (comparator.isLessOrEqual(x1, x0) && comparator.isLessOrEqual(x1, x2)) {
+            ranked = Ranked.of(x1, r1);
+            children = List.of(t0, t2).appendAll(t1.getChildren());
+        } else if (comparator.isLessOrEqual(x2, x0) && comparator.isLessOrEqual(x2, x1)) {
+            ranked = Ranked.of(x2, r2);
+            children = List.of(t0, t1).appendAll(t2.getChildren());
+        } else {
+            ranked = Ranked.of(x0, r1);
+            assert t0.getChildren().isEmpty(); // TODO Scalaz has this suffix, the original paper doesn't
+            children = List.of(t1, t2);
+        }
+        return Tree.of(ranked, children);
+    }
+
+
+    /**
+     * fun ins (t, []) = [t]
+     * * | ins (t, t' :: ts) = (∗ rank t ≤ rank t' ∗)
+     * *     if rank t < rank t' then t :: t' :: ts
+     * *     else                     ins (link (t, t'), ts)
+     */
+    private List<Node<Ranked<T>>> ins(Node<Ranked<T>> tree, List<Node<Ranked<T>>> forest) {
```

Comment:
not sure when to return `List<Node<Ranked<T>>` and when to wrap it in a `PriorityQueue`.

For now private methods (with intermediary stages) use the `List`, public methods accept and return `PriorityQueue`.